tests/Microsoft.Identity.Test.Unit/PublicApiTests/ConfidentialClientApplicationTests.cs

```diff
@@ -1672,6 +1672,48 @@ public async Task ValidateGetAccountAsyncWithNullEmptyAccountIdAsync(string acco
                 Assert.IsNull(acc);
             }
         }
+
+        [TestMethod]
+        public async Task ValidateAppTokenProviderAsync()
+        {
+            var app = ConfidentialClientApplicationBuilder.Create(TestConstants.ClientId)
+                                                          .WithAppTokenProvider(async (AppTokenProviderParameters parameters) =>
+                                                          {
+                                                              Assert.IsNotNull(parameters.Scopes);
+                                                              Assert.IsNotNull(parameters.CorrelationId);
+                                                              Assert.IsNotNull(parameters.TenantId);
+                                                              Assert.IsNotNull(parameters.CancellationToken);
+
+                                                              return Task.Run(() => GetAppTokenProviderResult()).Result;
+                                                          })
+                                                          .BuildConcrete();
+            await app.AcquireTokenForClient(TestConstants.s_scope).ExecuteAsync(new CancellationToken()).ConfigureAwait(false);
+
+            var tokens = app.AppTokenCacheInternal.Accessor.GetAllAccessTokens();
+
+            Assert.IsTrue(tokens.Count > 0);
+
+            var token = tokens.FirstOrDefault();
+            Assert.IsNotNull(token);
+            Assert.AreEqual(TestConstants.DefaultAccessToken, token.Secret);
+
+            var result = await app.AcquireTokenForClient(TestConstants.s_scope).ExecuteAsync(new CancellationToken()).ConfigureAwait(false);
+            
+            Assert.IsNotNull(result.AccessToken);
+            Assert.AreEqual(TestConstants.DefaultAccessToken, result.AccessToken);
+            Assert.AreEqual(TokenSource.Cache, result.AuthenticationResultMetadata.TokenSource);
+        }
+
+        private ExternalTokenResult GetAppTokenProviderResult()
+        {
+            ExternalTokenResult token = new ExternalTokenResult();
+            token.RawAccessToken = TestConstants.DefaultAccessToken;
+            token.CorrelationId = Guid.NewGuid().ToString();
+            token.ExpiresInSeconds = 3600 + DateTimeHelpers.CurrDateTimeInUnixTimestamp();
```

Comment:
Ah, now I understand why your product code works. I'm pretty sure that ESTS returns ExpiresIn as `3600` not as `12349013409123`. 

We only use UnixTimestamps when we store the values in the token cache. This is done for you at a different layer. You just need to send `3600`

Reply:
Ok, I mix up the fact that ESTS returns just the length of time when the token is valid rather than the point of time in which the token will expire.